src/config.js

```diff
@@ -13,41 +13,82 @@ const log = createLogger(__filename);
 type ApplyConfigToArgvParams = {|
   argv: Object,
   configObject: Object,
-  defaultValues: Object,
+  defaultValues?: Object,
+  subCommandDefaultValues: Object,
   configFileName: string,
+  commandExecuted?: string,
 |};
 
 export function applyConfigToArgv({
   argv,
   configObject,
   defaultValues = {},
+  subCommandDefaultValues,
   configFileName,
+  commandExecuted,
 }: ApplyConfigToArgvParams): Object {
-  const newArgv = {...argv};
+  let newArgv = {...argv};
   for (const option in configObject) {
-    // we assume the value was set on the CLI if the default value is
-    // not the same as that on the argv object as there is a very rare chance
-    // of this happening
-    if (camelCase(option) !== option) {
-      throw new UsageError(`The config option "${option}" must be ` +
-        `specified in camel case: "${camelCase(option)}"`);
-    }
-    const wasValueSetOnCLI = typeof(argv[option]) !== 'undefined' &&
-      (argv[option] !== defaultValues[option]);
-    if (wasValueSetOnCLI) {
-      log.debug(`Favoring CLI: ${option}=${argv[option]} over ` +
-        `configuration: ${option}=${configObject[option]}`);
+
+    validateOption(option);
+    if (option === commandExecuted) {
+      configObject = configObject[option];
+      newArgv = applyConfigToArgv({
+        argv,
+        configObject,
+        subCommandDefaultValues,
+        configFileName});
       continue;
     }
-    if (!argv.hasOwnProperty(decamelize(option, '-'))) {
-      throw new UsageError(`The config file at ${configFileName} specified ` +
-        `an unknown option: "${option}"`);
+
+    if (defaultValues[option] === undefined && subCommandDefaultValues) {
+      defaultValues = subCommandDefaultValues;
+    }
+
+    if (wasValueSetOnCLI(option, argv[option], defaultValues[option],
+      configObject[option])) {
+      continue;
     }
+
+    checkOptionPresence(argv, option, configFileName);
+
     newArgv[option] = configObject[option];
   }
   return newArgv;
 }
 
+function checkOptionPresence(argv: Object, option: string,
+  configFileName: string) {
+  if (!argv.hasOwnProperty(decamelize(option, '-'))) {
+    throw new UsageError(`The config file at ${configFileName} specified ` +
+      `an unknown option: "${option}"`);
+  }
+
+}
+
+function validateOption(option: string) {
+  if (camelCase(option) !== option) {
+    throw new UsageError(`The config option "${option}" must be ` +
+      `specified in camel case: "${camelCase(option)}"`);
+  }
+}
+
+function wasValueSetOnCLI(optionName: string, optionValue: string,
```

Comment:
I know this has a bit more code to it but I also don't think it deserves its own function (since it's not shared)